src/rand_impls.rs

```diff
@@ -269,9 +269,38 @@ mod tests {
 
     #[test]
     fn floating_point_edge_cases() {
-        // the test for exact equality is correct here.
-        assert!(ConstantRng(0xffff_ffff).gen::<f32>() != 1.0);
-        assert!(ConstantRng(0xffff_ffff_ffff_ffff).gen::<f64>() != 1.0);
+        const EPSILON32: f32 = 1.0 / (1u32 << 23) as f32;
+        const EPSILON64: f64 = 1.0 / (1u64 << 52) as f64;
+
+        let mut zeros = ConstantRng(0);
+        let mut ones = ConstantRng(!0);
+
+        let zero32 = zeros.gen::<f32>();
+        let zero64 = zeros.gen::<f64>();
+        let one32 = ones.gen::<f32>();
+        let one64 = ones.gen::<f64>();
+        assert_eq!(zero32, 0.0);
+        assert_eq!(zero64, 0.0);
+        assert!(1.0 - EPSILON32 <= one32 && one32 < 1.0);
+        assert!(1.0 - EPSILON64 <= one64 && one64 < 1.0);
```

Comment:
As noted above, I believe this should be exactly `1 - ε / 2`. Exact equality tests here would be preferable (excepting the `Open01` case near 0, where there are many near representable values).

Reply:
Actually, `one32` should be `1.0 - f32::EPSILON`, not `1.0 - f32::EPSILON / 2`, since the generator generates a multiple of ɛ in `gen::<f32>`. If we are to use the exact equality tests, they should be `(zero32, one32) == (0.0, 1.0 - f32::EPSILON)`, `(closed_zero32, closed_one32) == (0.0, 1.0)` and `(open_zero32, open_one32) == (f32::EPSILON / 2, 1.0 - f32::EPSILON / 2)`. Should I amend the commit to do that?